### PR TITLE
Mm 1359

### PR DESCRIPTION
**Pull request by ArdonToonstra**
- Verduidelijken beschrijvingen
- Verbeteren FHIRpath statements
- Toevoegen asserts
- Verwijderen invoer van datum T door de gebruiker. Dat is hier niet nodig.
- Ondersteuning voor zowel relatieve of absolute verwijzingen naar PDFA documenten of DocumentReferences

### CODE REVIEW

File: FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-4-serve-2-pdfa.xml [hunk 1]
```diff
@@ -39,13 +39,19 @@
         <description value="OAuth Token for patient XXX_Baltus"/>
     </variable>   
     <variable>
-        <name value="binary1-url"/>
-        <expression value="Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='68688-1')).content[0].attachment.url.substring(7)"/>
+        <name value="pdfa1-url"/>
+        <!-- Pick the absolute url as is or create an absolute url when a relative url is given by combining the base URL given in the Bundle self link with the relative reference.-->
+        <expression value="iif(Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='68688-1')).content[0].attachment.url.startsWith('Binary/'),
+            Bundle.link.url.replaceMatches('DocumentReference[/?].*$','') + Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='68688-1')).content[0].attachment.url,
+            Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='68688-1')).content[0].attachment.url)"/>
```
Comment (jduwel) [inline FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-4-serve-2-pdfa.xml:46]: Voor de leesbaarheid is het wel lekker om 
`Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='68688-1')).content[0].attachment.url`

in een losse variabele te definiëren.

File: FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-4-serve-2-pdfa.xml [hunk 2]
```diff
@@ -39,13 +39,19 @@
...
         <sourceId value="documentreference-search-response"/>
     </variable>
     <variable>
-        <name value="binary2-url"/>
-        <expression value="Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='34781-5')).content[0].attachment.url.substring(7)"/>
+        <name value="pdfa2-url"/>
+        <!-- Pick the absolute url as is or create an absolute url when a relative url is given by combining the base URL given in the Bundle self link with the relative reference.-->
+        <expression value="iif(Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='34781-5')).content[0].attachment.url.startsWith('Binary/'),
+            Bundle.link.url.replaceMatches('DocumentReference[/?].*$','') + Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='34781-5')).content[0].attachment.url,
+            Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='34781-5')).content[0].attachment.url)"/>
```
Comment (jduwel) [inline FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-4-serve-2-pdfa.xml:54]: Idem voor `Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code ='68688-1')).content[0].attachment.url`

**Comment by eduardderijcke** [comment]
@ArdonToonstra Could you may be check the feedback of Jorn? After your opinion I'll review this PR.

**Comment by ArdonToonstra** [comment]
Discussed with Jorn that we will leave it as is, with regards to the suggestions to use variables in the FHIRPath statements. The main reason is that this will be difficult to transfer to the NTS format as this will result in a variable in a variable which Touchstone does not support. 
Will work on putting this in the NTS format in the (near) future.

File: FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml [hunk 3]
```diff
@@ -96,39 +59,110 @@
...
             </assert>
-        </action>-->
+        </action>
         <action>
             <assert>
-                <description value="Confirm that the returned resource type is Bundle."/>
+                <description value="Confirm that the returned Bundle total value matches or is less than the number of entries in the Bundle. The included resources should not be counted as entries in the Bundle.total."/>
                 <direction value="response"/>
-                <resource value="Bundle"/>
+                <expression value="Bundle.total.toInteger() &lt;= Bundle.entry.where(search.empty() or search.mode = 'match').count()"/>
```
[the remark this answers is not in the document]
Comment (ArdonToonstra) [reply FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml:87]: Yes! :) It is a XML escaped character: https://www.freeformatter.com/xml-escape.html

File: FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml [hunk 4]
```diff
@@ -168,31 +202,31 @@
...
                 <warningOnly value="true"/>
             </assert>
         </action>
         <action>
             <assert>
-                <description value="Confirm that the returned DocumentReference does not have a similiar masterIdentifier as the corresponding fixture. Set to warningonly because our own testruns should fail on this assert"/>
+                <description value="Confirm that the returned DocumentReference does not have a similiar masterIdentifier as the corresponding fixture. Set to warningonly because our own testruns should fail on this assert."/>
                 <direction value="response"/>
-                <expression value="Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code = '34781-5')).masterIdentifier.value"/>
+                <expression value="Bundle.entry.select(resource as DocumentReference).where(type.coding.where(code = '34797-1')).masterIdentifier.value"/>
```
[the remark this answers is not in the document]
Comment (ArdonToonstra) [reply FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml:217]: Good one! I will double check and fix.

File: FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml [hunk 5]
```diff
@@ -96,39 +59,110 @@
...
+        <action>
+            <assert>
+                <description value="Confirm that literal References (Reference.reference) are an absolute URL, a relative URL or an internal fragment reference (contained), see: http://hl7.org/fhir/stu3/references.html#literal."/>
+                <direction value="response"/>
+                <expression value="Bundle.descendants().reference.all( startsWith('http://') or startsWith('https://') or startsWith('#') or matches('^urn:oid:[0-2](\\.(0|[1-9]\\d*))*$') or matches('^urn:uuid:[A-Fa-f\\d]{8}-[A-Fa-f\\d]{4}-[A-Fa-f\\d]{4}-[A-Fa-f\\d]{4}-[A-Fa-f\\d]{12}$') or (startsWith('urn:').not() and startsWith('#').not() and matches('^[A-Za-z]{3,}/[^/]+$')) )"/>
+            </assert>
+        </action>
+        <action>
+            <assert>
+                <description value="Confirm that all References have a display value, see https://informatiestandaarden.nictiz.nl/wiki/MedMij:V2019.01_FHIR_IG#Use_of_the_reference_datatype."/>
```
[the remark this answers is not in the document]
Comment (ArdonToonstra) [reply FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml:148]: Will do!

File: FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArdonToonstra) [reply FHIR3-0-2-MM202001-Dev/PDFA-3-0/XIS-Server/medmij-pdfa-fhir3-0-2-xis-1-1-serve-3-documentreference.xml:217]: Mm, I don't think that matters.

**Pull request merged**